### ISSUE TITLE
train real-world swinIR with the given running sentence

### ISSUE THREAD

**Issue by aiaini66** [cszn/KAIR#99]
Hello, when I was using  "#003 Real-World Image SR"  to train my own super-resolution model, I found this sentence in the KAIR/docs/README_SwinIR.md file: "# before training gan, put the PSNR-oriented model into superresolution /swinir_sr_realworld_x4_gan/models/", what is this for? Will there be any impact on training if you do not follow this operation? I did not find the PSNR-oriented model, and I did not find a place to use it in the training code.

**Comment by aiaini66** [comment]
@JingyunLiang

**Comment by JingyunLiang** [comment]
Training real-world image SR takes a two-step strategy: first train the PSNR-based model, then the GAN-based model. Therefore, you have to fine-tune from the pretrained PSNR-based model. All models are provided at https://github.com/JingyunLiang/SwinIR/releases.

If you want to train it from scratch, you should first train the PSNR-based model (You can still fine-tune from the pretrained PSNR-based model for quicker convergence).

**Comment by aiaini66** [comment]
Thank you for such a quick reply. I have another question. I haven't seen the PSNR model being used in  total train code?

**Comment by aiaini66** [comment]
@JingyunLiang

**Comment by JingyunLiang** [comment]
If you put the PSNR-oriented model into `superresolution /swinir_sr_realworld_x4_gan/models/`, it should load it automatically.

**Comment by aiaini66** [comment]
I have  known what's going on. This PSNR-oriented is actually these models  "003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_PSNR-with-dict-keys-params-and-params_ema.pth
 003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_PSNR.pth
 003_realSR_BSRGAN_DFO_s64w8_SwinIR-M_x4_PSNR-with-dict-keys-params-and-params_ema.pth
 003_realSR_BSRGAN_DFO_s64w8_SwinIR-M_x4_PSNR.pth", right?
 It corresponds to line in main_train_psnr.py
 “init_iter_G, init_path_G = option.find_last_checkpoint(opt['path']['models'], net_type ='G') “

**Comment by aiaini66** [comment]
@JingyunLiang

**Comment by JingyunLiang** [comment]
Correct.

**Comment by aiaini66** [comment]
![Screenshot from 2021-11-22 18-40-17](https://user-images.githubusercontent.com/27721212/142847247-b96f66a6-1180-48dc-b152-488e7126ffe6.png)


The last question about parameters: The img_size parameter is only used in the initialization of the SwinTransformerBlock class（As shown above ）. In actual testing and training, when the size of the input image is not equal to img_size, it is equivalent to not using this parameter. Is it? Is my understanding wrong?

**Comment by aiaini66** [comment]
@JingyunLiang  thanks a lot

**Comment by JingyunLiang** [comment]
We initialize the mask as `self.attn_mask` according to `self.input_resolution`, so that we don't need to calculate the mask for every iteration.  Note that the training image size is fixed, e.g., 64x64.

In testing, we calculate the mask for a given testing image. Note that the testing image is generally not 64x64.

**Comment by aiaini66** [comment]
I trained my own data with the training code of swinIR (not loaded your pre-trained model) some time ago. The configuration is the default in KAIR. The file used is options/swinir/train_swinir_sr_realworld_x4_gan.json, but the effect is worse than that of the pre-trained model. Poor, what could be the reason for this? I have been looking for the reason for a long time, and I eagerly hope for your suggestions.  
there are source_img   your_pre-trained_model_result   my_trained_model_result
![Screenshot from 2021-11-22 18-51-43](https://user-images.githubusercontent.com/27721212/142849747-bcd3a581-633d-4396-93ea-1944833ebf37.png)
@JingyunLiang

**Comment by aiaini66** [comment]
> 

oh, i got it thanks!

**Comment by aiaini66** [comment]
@JingyunLiang  
I’m really sorry, I’m here to bother you again. I found that if I loaded the pre-trained model you gave to train my own data, The effect of the pre-trained model can be achieved during the test, but if I don't loaded the pre-trained model to train my own data,  the test effect is worse than the pre-trained model.  Do you have any other strategies when you train? Such as degenerate operation, parameter setting and so on. . .

**Comment by JingyunLiang** [comment]
No other tricks. Do you train enough time and use the same training data?

**Comment by aiaini66** [comment]
I use the same training data as your , and have train for 161,701 iters, batch size is 5.  The current result is still smoother than your trained, and it looks a lot blurry. Is this because of insufficient training time? @JingyunLiang

**Comment by JingyunLiang** [comment]
Do you use GAN training? You first need to pretrain it with L1 loss (blurry results but high PSNR). Then, you train it with a combination of L1loss + GAN loss + perceptual loss (visually pleasant images with lower PSNR).

**Comment by aiaini66** [comment]
I used GAN training, but I didn't just use L1 loss training first. I used L1loss + GAN loss + perceptual loss throughout the training process, and the weights are also the default, which are 1, 1, 0.1. Some of the parameters are shown in the following figure:
![Screenshot from 2021-11-24 14-11-32](https://user-images.githubusercontent.com/27721212/143184829-8310e171-b39d-4f86-8724-76c52a96f2c8.png)


**Maybe this is one reason why the result is not good !    Can you tell me how much data is iterated before adding GAN loss and perceptual loss? and What are their respective weights?   @JingyunLiang

**Comment by JingyunLiang** [comment]
First use https://github.com/cszn/KAIR/blob/master/options/swinir/train_swinir_sr_realworld_x4_psnr.json then use https://github.com/cszn/KAIR/blob/master/options/swinir/train_swinir_sr_realworld_x4_gan.json. GAN training should use PSNR-oriented pretrained model.

**Comment by aiaini66** [comment]
I will train according to your suggestion first, and then synchronize the training result  with you later, thank you very much!

**Comment by JingyunLiang** [comment]
It is recommended to read the [Readme](https://github.com/cszn/KAIR/blob/master/docs/README_SwinIR.md) carefully.

**Comment by aiaini66** [comment]
I want to train this large model “”“003_realSR_BSRGAN_DFOWMFC_s64w8_SwinIR-L_x4_GAN.pth”, do I only need to change these four parameters  corresponding to the large model, and nothing else needs to be changed?
there are:
   , "depths": [6, 6, 6, 6, 6, 6, 6, 6, 6]
     , "embed_dim": 240
     , "num_heads": [8, 8, 8, 8, 8, 8, 8, 8, 8, 8]
     , "resi_connection": "3conv"
@JingyunLiang

**Comment by JingyunLiang** [comment]
+ the training datasets

**Comment by aiaini66** [comment]
> 

ok, thanks !

**Comment by sarvghotra** [comment]
@JingyunLiang  For large SwinIR model, is there any change for netD?

**Comment by JingyunLiang** [comment]
1, No
2, netD starts from scratch.

**Comment by sarvghotra** [comment]
@JingyunLiang 

I tried one complete run with phase 1 and 2 training, but couldn't get good enough results. So, have another question before trying again: for phase 2 training, GAN and other losses, is the G net's optimizer started from scratch or per-loaded from phase 1 training? I am confused due to this: https://github.com/cszn/KAIR/blob/72e93351bca41d1b1f6a4c3e1957f5bffccc7101/options/swinir/train_swinir_sr_realworld_x4_gan.json#L106

Thanks!
